pkg/rpc/signal_interceptor.go

```diff
@@ -0,0 +1,31 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpc
+
+import (
+	"context"
+	"os/signal"
+	"syscall"
+
+	"google.golang.org/grpc"
+)
+
+// SignalHandlingInterceptor is a gRPC interceptor that cancels the context when a termination signal is received.
+func SignalHandlingInterceptor(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp any, err error) {
+	ctx, cancel := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	return handler(ctx, req)
+}
```

Comment:
@Warashi Sorry, I noticed that we can't distinguish Graceful Shutdown and context cancel for now. piped still treats both as context canceled.

We might use `signal.Notify` like this.
```suggestion
// SignalHandlingInterceptor is a gRPC interceptor that cancels the context when a termination signal is received.
var terminated atomic.Bool

// SignalHandlingInterceptor is a gRPC interceptor that cancels the context when a termination signal is received.
func SignalHandlingInterceptor(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp any, err error) {
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	go func() {
		select {
		case <-sigCh:
			// received a termination signal
			terminated.Store(true)
			cancel()
		}
	}()

	return handler(ctx, req)
}
```

Reply:
@ffjlabo 
I implemented the determination of the Graceful Shutdown in this file. Is it not enough?

[`pkg/plugin/signalhandler/handler.go`](https://github.com/pipe-cd/pipecd/pull/5390/files#diff-97d64f7b9d2af9a5ee8ac1e6ded6c7313aba3c14ecae6cdd0596f044c3d962b6)

Reply:
@Warashi  I got it. Sorry, I missed it. 🙏